go/vt/key/key.go

```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
 	return buf
 }
 
-//
-// KeyRange helper methods
-//
+// Helper methods for keyspace id values.
 
-// EvenShardsKeyRange returns a key range definition for a shard at index "i",
-// assuming range based sharding with "n" equal-width shards in total.
-// i starts at 0.
-//
-// Example: (1, 2) returns the second out of two shards in total i.e. "80-".
-//
-// This function must not be used in the Vitess code base because Vitess also
-// supports shards with different widths. In that case, the output of this
-// function would be wrong.
-//
-// Note: start and end values have trailing zero bytes omitted.
-// For example, "80-" has only the first byte (0x80) set.
-// We do this to produce the same KeyRange objects as ParseKeyRangeParts() does.
-// Because it's using the Go hex methods, it's omitting trailing zero bytes as
-// well.
-func EvenShardsKeyRange(i, n int) (*topodatapb.KeyRange, error) {
-	if n <= 0 {
-		return nil, fmt.Errorf("the shard count must be > 0: %v", n)
-	}
-	if i >= n {
-		return nil, fmt.Errorf("the index of the shard must be less than the total number of shards: %v < %v", i, n)
-	}
-	if n&(n-1) != 0 {
-		return nil, fmt.Errorf("the shard count must be a power of two: %v", n)
+// Normalize adds padding to id as necessary to make sure it can be compared as an 8 byte integer.
+// From a key range perspective, -80 == 00-80 == 0000-8000 == 000000-800000, etc. and they should
+// always be treated the same even if they are different lengths.
+func Normalize(id []byte) []byte {
+	if len(id) >= 8 {
+		return id[:8]
```

Comment:
One breaking change that we should talk about is restricting it is 8-byte length. which the current implementation does not do.
Current implementation support keyspace id to be arbitrary length based on the Vindex.

Reply:
An example of longer-than-8-byte keyspace IDs can be seen [here](https://vitess.io/docs/16.0/concepts/shard/#shard-naming) (link courtesy @harshit-gangal)

Reply:
Yeah, I (also) mistakenly thought they were always 64 bits, but that's only the length of the default `hash` vindex based keyspace IDs. Some are shorter and others longer. You can see examples of how many hex digits are produced for each vindex type by using the vindex query functions: https://vitess.io/docs/17.0/reference/features/vindexes/#query-vindex-functions 

I was also pointed this blog post that I don't remember previously reading: https://vitess.io/blog/2017-09-18-custom-sharding-with-vitess/

Reply:
I see what you mean here. I am not sure that's a real case that anyone is using but I am quite okay with not changing the semantics here. We can pivot the `Normalize` method to, instead of padding out to a fixed length (which there isn't any value that would satisfy the requirements without defining a maximum length) to _instead_ remove all trailing zeroes consistently.

I made that change locally and it works fine and all tests pass the same except for `TestNormalize` (since it's explicitly checking the return values of `Normalize`, so it would be expected). I'll push another commit for this.